### PR TITLE
53 - Add Login / Logout

### LINKED ISSUE #53: Add Login / Logout mechanism
## Overview of the Feature Request

To be able to use the SPA as an authenticated user, it will be necessary to first log in via JSF login page, which will setup the JSESSIONID cookie, necessary for authentication (just for the MVP, see: [strategy doc](https://docs.google.com/document/d/1vOSsIGGCbgal4f_lSp1yInQifBcmo5vZ-Z9pfkVDyG4)).

The goal is to add a Log In button which redirects to the JSF Log In page and a Log Out mechanism that removes the JSESSIONID cookie.

## What kind of user is the feature intended for?

Test users.

## What inspired the request?

- @pdurbin 
- Better UX for the SPA, than having to manually navigate to JSF login

## What existing behavior do you want changed?

No

## Any brand new behavior do you want to add to Dataverse?

Add a Login / Logout mechanism to the SPA.

## Any open or closed issues related to this feature request?

- https://github.com/IQSS/dataverse/issues/9531
- https://github.com/IQSS/dataverse-client-javascript/issues/56

### PR DESCRIPTION
**Pull request by GPortas**
## What this PR does / why we need it:

- This PR must be merged first (to resolve the conflicts in this PR): https://github.com/IQSS/dataverse-frontend/pull/54

Adds the Login / Logout feature to the header component.

Log In and Sign Up redirects to JSF. To get the current authenticated user and Log Out, the application calls the js-dataverse package, which wraps those API operations.

## Which issue(s) this PR closes:

- Closes #53 

## Special notes for your reviewer:

This PR is part of an end-to-end flow that covers three PRs from three different repositories:
- https://github.com/IQSS/dataverse/pull/9533
- https://github.com/IQSS/dataverse-client-javascript/pull/57

This branch includes a copy of the js-dataverse module with the latest changes (js-dataverse-2.0.0.tgz). As the module has not currently been released and therefore has not been pushed with these changes to npm either, it is the fastest way to have it available for development.

### New GitHub environment variables required

The SPA now requires an .env file in order to be built, where the environment variable DATAVERSE_BACKEND_URL is specified. There are different points at which the SPA is built through GitHub actions and therefore require the setup of new secrets for such variable.

It is important that the variable is specified at the time of building, otherwise the application will fail once deployed. See: https://github.com/IQSS/dataverse-frontend/blob/e2aaddbb0a0bf92a67186ac5ee6109b33d38349e/src/App.tsx#L8

#### e2e test action

This action requires that a new environment called "e2e" is created with a DATAVERSE_BACKEND_URL secret. At the moment, since e2e do not execute API calls, we can place a random URL. In the future we will have to specify a URL of a real instance for e2e testing.

#### deploy action

Within the "qa" environment, the DATAVERSE_BACKEND_URL secret must be also added, with the URL corresponding to the Dataverse instance to be used for QA. Keep in mind that to make API requests work the instance must be deployed with the changes developed in: `GPortas:9531-session-api-auth-cors`

## Suggestions on how to test this:

Clone Dataverse repository pointing to branch `GPortas:9531-session-api-auth-cors` (https://github.com/IQSS/dataverse/pull/9533) 

Before building and deploying the Dataverse application locally, add the following properties to the microprofile-config.properties file:

`
dataverse.feature.api-session-auth=true
`

Once Dataverse is deployed and running on `localhost:8080`, run the SPA in development mode:

`npm start`

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:

logout_1

### CODE REVIEW

**Comment by GPortas** [comment]
The e2e action is failing as the "e2e" environment and its DATAVERSE_BACKEND_URL secret are not defined in the repository configuration.

**Comment by pdurbin** [comment]
@GPortas when you get a chance, can you please resolve the merge conflicts in this pull request? Thanks! ❤️

File: package.json [hunk 1]
```diff
@@ -104,12 +106,14 @@
     "eslint-plugin-storybook": "^0.6.11",
     "eslint-plugin-unused-imports": "^2.0.0",
     "eslint-plugin-vitest": "^0.0.53",
+    "js-dataverse": "file:./js-dataverse-2.0.0.tgz",
```
Comment (pdurbin) [inline package.json:109]: If we do this for years it will add a lot of bloat to the code. I started a Slack thread to discuss this: https://iqss.slack.com/archives/C04MJ76A34Z/p1682594807564159

**Review by pdurbin — COMMENTED** [review]
For now, just a quick comment about the potential for bloat in the long term.

File: package.json (hunk 1 above)
Comment (pdurbin) [reply package.json:109]: @GPortas thanks for creating this issue about it!

- #77

I'm not worried about this ~10KB file. We'll fix this in the future. Resolving.

File: src/sections/route.enum.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 export enum Route {
   HOME = '/',
-  SIGN_UP = '/',
-  LOG_IN = '/',
+  SIGN_UP = '/dataverseuser.xhtml?editMode=CREATE',
+  LOG_IN = '/loginpage.xhtml',
```
Comment (pdurbin) [inline src/sections/route.enum.ts:4]: We want to try something like

```
loginpage.xhtml?redirectPage=%2Fdataverse.xhtml
```

**Review by pdurbin — COMMENTED** [review]
Works in Chrome, not Firefox.

After logging in via JSF, it should redirect. Currently, you are stuck after clicking Login even though it worked.

A couple screenshots from Firefox:

Screen Shot 2023-04-27 at 12 28 37 PM
Screen Shot 2023-04-27 at 12 33 47 PM

**Comment by GPortas** [comment]
I have updated js-dataverse module including `'Content-Type': 'application/json'` for post requests.

The issue no longer appears when testing on my remote test instance with Firefox 112.0.2.

@pdurbin Please let me know if that fixes your environment as well.

**Comment by pdurbin** [comment]
@GPortas can you please resolved the merge conflicts in this PR?

**Comment by GPortas** [comment]
@pdurbin I just resolved the conflicts by updating with develop.

That animated gif is great! Although I think you're not testing something important, which is executing the log out from the SPA, not from JSF. That was the trigger for the issue we detected in Firefox.

Please, can you test that SPA log out works?

Thanks!

**Comment by pdurbin** [comment]
@GPortas sure, let me try again. Why is e2e push failing?

**Comment by pdurbin** [comment]
@GPortas 

I'm not sure why but now I can't seem to log into the SPA at all. I built the frontend war file from ba8c085 and here it is if you'd like to take a look: [dataverse-frontend.war.zip](https://github.com/IQSS/dataverse-frontend/files/11421630/dataverse-frontend.war.zip)

A gif of the login attempts:

login-logout2

**Comment by GPortas** [comment]
> The e2e action is failing as the "e2e" environment and its DATAVERSE_BACKEND_URL secret are not defined in the repository configuration.

Hey @pdurbin, it will fail until we set the GitHub environment and a secret value. Thanks!

**Comment by pdurbin** [comment]
@GPortas right, right, I think we discussed the e2e test and I forgot to circle back and add the secrets for you. I'll bug you on Slack about this.

Meanwhile, I rebuilt the war file (stopped Storybook, removed the "dist" directory), and it works fine now. Sorry for the noise! Here's a gif showing how in Firefox you can logout from the SPA now:

login-logout3

I'm going to go ahead an move the whole lot (all three PRs) to QA. Thanks!

**Comment by kcondon** [comment]
Issues found:
1. When signing up from SPA, first attempt appears to do nothing, remains on signup page but throws server log null ptr. Clicking create again sends you to mydata page but logged in. If rather than clicking create again you went to homepage you would see the account was created successfully. See attached log for error:
[signup_err.txt](https://github.com/IQSS/dataverse-frontend/files/11444521/signup_err.txt)

2. When refreshing the SPA homepage if not logged in, see a server log error. Note it also happens if you just type the url in the browser without hitting enter, I believe because it attempts to prefetch site icon? See attached log error:
[server_log_err_not_logged_in_spa.txt](https://github.com/IQSS/dataverse-frontend/files/11444528/server_log_err_not_logged_in_spa.txt)
